src/charm.py

```diff
@@ -57,17 +58,36 @@ def __init__(self, *args: typing.Any):
         self.framework.observe(
             self.on[constants.PEER].relation_departed, self._on_peer_relation_departed
         )
-        self.peer_relation = self.model.get_relation(constants.PEER)
+        self.framework.observe(
+            self.on[constants.PEER].relation_joined, self._on_peer_relation_joined
+        )
 
     def _on_reload_bind(self, _: events.ReloadBindEvent) -> None:
         """Handle periodic reload bind event."""
+        # Reloading is used to take new configuration into account
         try:
             relation_data = self.dns_record.get_remote_relation_data()
         except ValueError as err:
             logger.info("Validation error of the relation data: %s", err)
             return
-        if self.bind.has_a_zone_changed(relation_data):
-            self.bind.update_zonefiles_and_reload(relation_data)
+        topology = self._topology()
+        if self.bind.has_a_zone_changed(relation_data, topology):
+            self.bind.update_zonefiles_and_reload(relation_data, topology)
+
+    def _on_peer_relation_joined(self, _: ops.RelationJoinedEvent) -> None:
+        """Handle peer relation joined event."""
+        # Reloading is used to take new configuration into account
+        try:
+            relation_data = self.dns_record.get_remote_relation_data()
+        except ValueError as err:
+            logger.info("Validation error of the relation data: %s", err)
+            return
+        try:
+            topology = self._topology()
+            self.bind.update_zonefiles_and_reload(relation_data, topology)
+        except (PeerRelationUnavailableError, PeerRelationNetworkUnavailableError) as err:
+            logger.info("Could not retrieve network topology: %s", err)
+            return
```

Comment:
Maybe we should retry this until the peer network becomes available? either defer or an error state retry.

Reply:
This should be unavailable only at the beginning (during install) and should put the charm in a waiting status.  
If we don't have that later on, then something has really gone wrong and having an error state is fine by me

Reply:
Does this mean that there will be guaranteed events happening after the `peer-relation-joined` if this one fails?

Reply:
The custom `reload-bind` event is dispatched every minute on each unit and it check/adjust the network topology as the `peer-relation-joined` event would do

Reply:
Gotcha, thanks for the explanation!